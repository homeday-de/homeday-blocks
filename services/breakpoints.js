let breakpoints = {};

export const matchMediaAvailable = (
  typeof window !== 'undefined'
  && typeof window.matchMedia === 'function'
);

export const mediaMatches = (breakpoint, {
  matchMedia = matchMediaAvailable ? window.matchMedia : () => false,
} = {}) => {
  if (typeof breakpoints[breakpoint] === 'undefined') {
    // Breakpoint not defined, we treat it as a media query string
    return matchMedia(breakpoint).matches;
  }

  return matchMedia(breakpoints[breakpoint]).matches;
};

export const setBreakpoints = (newBreakpoints) => {
  breakpoints = newBreakpoints;
};

export const getBreakpoints = () => breakpoints;

export default {
  matchMediaAvailable,
  mediaMatches,
  setBreakpoints,
  getBreakpoints,
};
