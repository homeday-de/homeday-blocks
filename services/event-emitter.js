import Vue from 'vue';

const EventEmitter = new Vue();

export default EventEmitter;
