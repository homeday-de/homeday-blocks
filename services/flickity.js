export function hidePaginationWhenNotNeeded() {
  const viewport = this.size.width;
  const fullSize = Math.round(this.slideableWidth);

  if (fullSize > viewport) {
    this.bindDrag();
    this.element.classList.remove('flickity--no-controls');
  } else if (fullSize <= viewport) {
    this.unbindDrag();
    this.element.classList.add('flickity--no-controls');
  }
}

export default {
  hidePaginationWhenNotNeeded,
};
