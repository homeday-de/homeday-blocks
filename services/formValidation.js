export function email(string) {
  // eslint-disable-next-line
  const regex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return regex.test(string.toLowerCase());
}

export function date(string) {
  // eslint-disable-next-line
  const regex = /[0-9]{4}-[0-9]{2}-[0-9]{2}/;
  return regex.test(string);
}

export default {
  email,
  date,
};
