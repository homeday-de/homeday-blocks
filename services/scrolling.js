import naturalScroll from 'natural-scroll';
import {
  disableBodyScroll,
  enableBodyScroll,
} from 'body-scroll-lock';

export function elOffset(el) {
  const box = el.getBoundingClientRect();
  const clientTop = document.clientTop || document.body.clientTop || 0;
  const clientLeft = document.clientLeft || document.body.clientLeft || 0;
  const scrollTop = window.pageYOffset || document.scrollTop || 0;
  const scrollLeft = window.pageXOffset || document.scrollLeft || 0;

  return {
    top: box.top + (scrollTop - clientTop),
    left: box.left + (scrollLeft - clientLeft),
    width: box.width,
    height: box.height,
  };
}

export function getDocumentHeight() {
  return Math.max(
    document.body.scrollHeight, document.documentElement.scrollHeight,
    document.body.offsetHeight, document.documentElement.offsetHeight,
    document.body.clientHeight, document.documentElement.clientHeight,
  );
}

export function isWindowFullyScrolled() {
  if (window.pageYOffset + document.documentElement.clientHeight >= getDocumentHeight()) {
    return true;
  }

  return false;
}

export function freezeScrolling(el) {
  // eslint-disable-next-line no-param-reassign
  el.style['-webkit-overflow-scrolling'] = 'touch';
  disableBodyScroll(el);
}

export function unfreezeScrolling(el) {
  // eslint-disable-next-line no-param-reassign
  el.style['-webkit-overflow-scrolling'] = 'auto';
  enableBodyScroll(el);
}

export function isElVisible({ el, offset = 0 }) {
  if (el == null) {
    return false;
  }

  const { top: offsetTop } = elOffset(el);

  if (offsetTop < window.pageYOffset + offset) {
    return false;
  }

  if (offsetTop > window.pageYOffset + window.innerHeight) {
    return false;
  }

  return true;
}

export function isScrolledPastElBottom({
  el,
  // The default offset will check whether we reached the bottom of the element
  // with the bottom edge of the screen
  offset = window.innerHeight,
}) {
  if (el == null) {
    return false;
  }

  const { top: offsetTop } = elOffset(el);

  const elBottom = offsetTop + el.clientHeight;

  if (elBottom > window.pageYOffset + offset) {
    return false;
  }

  return true;
}

export function scrollToEl({ el, offset = 0, onlyIfNotVisible = true }) {
  const toScroll = [
    document.documentElement,
    document.body,
  ];

  if (el == null || (onlyIfNotVisible && isElVisible({ el, offset }))) {
    return;
  }

  toScroll.forEach((toScrollEl) => {
    naturalScroll.scrollTop(toScrollEl, elOffset(el).top - offset);
  });
}

export default {
  elOffset,
  getDocumentHeight,
  isWindowFullyScrolled,
  freezeScrolling,
  unfreezeScrolling,
  isElVisible,
  isScrolledPastElBottom,
  scrollToEl,
};
