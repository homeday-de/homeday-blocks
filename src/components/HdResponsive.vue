<script>
import onResize from 'homeday-blocks/src/services/on-resize';
import {
  matchMediaAvailable,
  mediaMatches,
  getBreakpoints,
} from 'homeday-blocks/src/services/breakpoints';

export default {
  name: 'HdResponsive',
  props: {
    breakpoints: {
      type: Object,
      default: () => ({}),
    },
  },
  data() {
    return {
      matches: {},
      indeterminate: !matchMediaAvailable,
    };
  },
  watch: {
    breakpoints() {
      this.setResponsiveFlags();
    },
  },
  created() {
    this.setResponsiveFlags();
  },
  mounted() {
    this.addResizeEvents();
  },
  beforeDestroy() {
    this.removeResizeEvents();
  },
  methods: {
    setResponsiveFlags() {
      const breakpoints = {
        ...getBreakpoints(),
        ...this.breakpoints,
      };

      this.matches = Object.keys(breakpoints).reduce(
        (matches, breakpointKey) => ({
          ...matches,
          [breakpointKey]: mediaMatches(breakpoints[breakpointKey]),
        }),
        {},
      );
    },
    addResizeEvents() {
      onResize.onThrottled(this.setResponsiveFlags);
    },
    removeResizeEvents() {
      onResize.offThrottled(this.setResponsiveFlags);
    },
  },
  render() {
    if (typeof this.$scopedSlots.default === 'undefined') {
      return null;
    }

    return this.$scopedSlots.default({
      matches: this.matches,
      indeterminate: this.indeterminate,
    });
  },
};
</script>
