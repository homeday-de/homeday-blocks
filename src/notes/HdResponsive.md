By default, `HdResponsive` component relies on global breakpoints set by the [Breakpoints service](https://github.com/homeday-de/homeday-blocks/tree/master/src/services#breakpoints). It also supports the `breakpoints` prop which will be merged with the global breakpoints.