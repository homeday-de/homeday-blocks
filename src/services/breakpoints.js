import sassVariables from 'homeday-blocks/src/styles/variablesForJS.scss';

let breakpoints = {
  mobile: `(max-width: ${sassVariables.breakMobile - 1}px)`,
  tablet: `(min-width: ${sassVariables.breakMobile}px)`,
  desktop: `(min-width: ${sassVariables.breakTablet}px)`,
  desktopWide: `(min-width: ${sassVariables.breakDesktop}px)`,
  desktopExtraWide: `(min-width: ${sassVariables.breakDesktopWide}px)`,
};

export const matchMediaAvailable = window && typeof window.matchMedia === 'function';

const defaultMatchMedia = matchMediaAvailable ? window.matchMedia : () => false;

export function mediaMatches(
  breakpoint,
  { matchMedia = defaultMatchMedia } = {},
) {
  if (typeof breakpoints[breakpoint] === 'undefined') {
    // Breakpoint not defined, we treat it as a media query string
    return matchMedia(breakpoint).matches;
  }

  return matchMedia(breakpoints[breakpoint]).matches;
}

export function setBreakpoints(newBreakpoints) {
  breakpoints = newBreakpoints;
}

export function getBreakpoints() {
  return breakpoints;
}

export default {
  matchMediaAvailable,
  mediaMatches,
  setBreakpoints,
  getBreakpoints,
};
