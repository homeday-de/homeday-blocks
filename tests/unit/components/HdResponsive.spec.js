import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import { setBreakpoints } from '@/services/breakpoints';
import HdResponsive from '@/components/HdResponsive.vue';

const BREAKPOINTS = {
  s: '(max-width:599px)',
  m: '(min-width:600px)',
  l: '(min-width:900px)',
  xl: '(min-width:1200px)',
};

// https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const wrapperBuilder = wrapperFactoryBuilder(HdResponsive);

describe('HdResponsive', () => {
  afterEach(() => {
    // We reset the breakpoints after each test
    setBreakpoints({});
  });

  it('renders no markup', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('provides "matches" object as part of a scoped slot', () => {
    setBreakpoints(BREAKPOINTS);

    const wrapper = wrapperBuilder({
      scopedSlots: {
        default: '<div>{{Object.keys(props.matches)}}</div>',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('provides "indeterminate" flag as part of a scoped slot', () => {
    setBreakpoints(BREAKPOINTS);

    const wrapper = wrapperBuilder({
      scopedSlots: {
        default: '<div>{{props.indeterminate}}</div>',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('breakpoint changes are managed', () => {
    setBreakpoints(BREAKPOINTS);

    const mockedSetListeners = jest.fn();
    const wrapper = wrapperBuilder({
      scopedSlots: {
        default: `
          <div>I'm a Slot</div>
        `,
      },
      methods: {
        setListeners: mockedSetListeners,
      },
    });

    const modifiedBreakpoints = BREAKPOINTS;
    delete modifiedBreakpoints.s;

    wrapper.setProps({ breakpoints: modifiedBreakpoints });

    expect(mockedSetListeners).toHaveBeenCalled();
  });
});
