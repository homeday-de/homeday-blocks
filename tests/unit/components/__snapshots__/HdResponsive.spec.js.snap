// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdResponsive provides "indeterminate" flag as part of a scoped slot 1`] = `<div>true</div>`;

exports[`HdResponsive provides "matches" object as part of a scoped slot 1`] = `
<div>[
  "s",
  "m",
  "l",
  "xl"
  ]</div>
`;

exports[`HdResponsive renders no markup 1`] = `""`;
